pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java

```diff
@@ -193,69 +191,74 @@ public NettyChannelInboundHandler(RequestHandler handler, NettyServerMetrics met
       this(handler, metric, 100);
     }
 
-    public enum State {
-      INIT,
-      REQUEST_RECEIVED,
-      RESPONSE_WRITTEN,
-      RESPONSE_SENT,
-      EXCEPTION
-    }
-
-    /**
-     * Server Channel Handler State
-     */
-    private State _state = State.INIT;
-
     @Override
     public void channelRead(ChannelHandlerContext ctx, Object msg) {
-      _requestStartTime = System.currentTimeMillis();
+      final long requestStartTime = System.currentTimeMillis();
       LOGGER.debug("Request received by server !!");
-      _state = State.REQUEST_RECEIVED;
-      ByteBuf request = (ByteBuf) msg;
-      _lastRequsetSizeInBytes = request.readableBytes();
+
+      final ByteBuf request = (ByteBuf) msg;
+      final long requestSizeInBytes = request.readableBytes();
 
       //Call processing handler
-      _lastProcessingLatency = MetricsHelper.startTimer();
-      byte[] response = _handler.processRequest(ctx, request);
-      _lastProcessingLatency.stop();
-
-      // Send Response
-      ByteBuf responseBuf = Unpooled.wrappedBuffer(response);
-      _lastSendResponseLatency = MetricsHelper.startTimer();
-      ChannelFuture f = ctx.writeAndFlush(responseBuf);
-      _state = State.RESPONSE_WRITTEN;
-      f.addListener(this);
-      request.release();
+      final TimerContext requestProcessingLatency = MetricsHelper.startTimer();
+      final ChannelHandlerContext requestChannelHandlerContext = ctx;
+      ListenableFuture<byte[]> serializedQueryResponse = _handler.processRequest(ctx, request);
+      Futures.addCallback(serializedQueryResponse, new FutureCallback<byte[]>() {
+        void sendResponse(@Nonnull final byte[] result) {
+          requestProcessingLatency.stop();
+
+          // Send Response
+          final ByteBuf responseBuf = Unpooled.wrappedBuffer(result);
+          final TimerContext responseSendLatency = MetricsHelper.startTimer();
+          ChannelFuture f = requestChannelHandlerContext.writeAndFlush(responseBuf);
+          f.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture future)
+                throws Exception {
+              LOGGER.debug("Response has been sent !!");
+              responseSendLatency.stop();
+              _metric.addServingStats(requestSizeInBytes, result.length, 1L, false,
+                  requestProcessingLatency.getLatencyMs(), responseSendLatency.getLatencyMs());
+              long totalQueryTime = System.currentTimeMillis() - requestStartTime;
+              if (totalQueryTime > _defaultLargeQueryLatencyMs) {
+                LOGGER.info("Trace Info: request handler processing time : {}, send response latency: {}, total time to handle request: {}",
```

Comment:
Now is a good time to change "Trace Info" to "Slow query" or something like that.